src/com/mysmartlogon/gidsApplet/GidsPINManager.java

```diff
@@ -84,39 +92,199 @@ private GidsPIN GetPINByReference(byte reference) throws NotFoundException {
+     * @return true if operation was successful, false otherwise
+     */
+    public boolean CheckAdminAuthenticationState (final byte requiredState) {
+        final byte authState = status[0];
+
+        // Check for no current authentication
+        if (requiredState == ADMIN_NOT_AUTHENTICATED) {
+            return authState == ADMIN_NOT_AUTHENTICATED;
+        }
+
+        // Check if admin authentication state is contained within required state(s)
+        if ((byte) (authState & requiredState) != authState) {
+            return false;
+        }
+
+        // Check if admin authentication state matches any known states exactly (trying to detect memory corruption)
+        switch (authState) {
+            case EXTERNAL_CHALLENGE:
+            case EXTERNAL_AUTHENTICATED:
+            case MUTUAL_CHALLENGE:
+            case MUTUAL_AUTHENTICATED:
+                return true;
+            default: // No known state matched! Something must be going wrong with the variable.
+                return false;
+        }
+    }
+
+    /**
+     * Checks if the current application state matches any of the required state(s).
+     * @param requiredState One or several required state(s), to give several possible states use the OR operation
+     * @return true if operation was successful, false otherwise
+     */
+    public boolean CheckApplicationState (final byte requiredState) {
+        // Check if application state is contained within required state(s)
+        if ((byte) (applicationState & requiredState) != applicationState) {
+            return false;
+        }
+
+        // Check if application state matches any known states exactly (trying to detect memory corruption)
+        switch (applicationState) {
+            case CREATION_STATE:
+            case INITIALIZATION_STATE:
+            case OPERATIONAL_STATE:
+            case TERMINATION_STATE:
+                return true;
+            default: // No known state matched! Something must be going wrong with the variable.
+                return false;
+        }
+    }
+
+    /**
+     * Changes the current external/mutual authentication state. Only used internally.
+     * @param nextState Next state of the application
+     * @return true if operation was successful, false otherwise
+     */
+    private boolean SetAdminAuthenticationState (final byte nextState) {
+        ClearChallengeData(); // Avoiding replay attacks
+
+        // No further checks for deauthentication
+        if (nextState == ADMIN_NOT_AUTHENTICATED) {
+            status[0] = ADMIN_NOT_AUTHENTICATED;
+            return true;
+        }
+
+        if (nextState == EXTERNAL_CHALLENGE) {
+            if (CheckAdminAuthenticationState(ADMIN_NOT_AUTHENTICATED)) {
+                status[0] = EXTERNAL_CHALLENGE;
+                return true;
+            } else {
+                status[0] = ADMIN_NOT_AUTHENTICATED;
```

Comment:
Does the code forbid to generate as many challenges as you want without blocking the card ? My understanding of the code is that 1 challenge = success 2 challenge = failure 3 challenge = success 4 challenge = failure etc
See 12.6	GENERAL AUTHENTICATE of the GIDS application which do not forbid as many authentication